File: ruby/20.valid-parentheses.rb
# frozen_string_literal: true

BRACKET_PAIRS = {
  '(' => ')',
  '[' => ']',
  '{' => '}'
}.freeze

# @param {String} s
# @return {Boolean}
def is_valid(s)
  brackets = s.chars.each_with_object([]) do |c, stack|
    if BRACKET_PAIRS.key?(c)
      stack.push(c)
    elsif BRACKET_PAIRS[stack.pop] != c
      return false
    end
  end

  brackets.empty?
end

File: rust/src/20.valid-parentheses.rs
use crate::Solution;

#[allow(dead_code)]
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = Vec::new();

        for char in s.chars() {
            match char {
                '(' => stack.push(')'),
                '[' => stack.push(']'),
                '{' => stack.push('}'),
                ')' | ']' | '}' if Some(char) != stack.pop() => return false,
                _ => (),
            }
        }

        stack.is_empty()
    }
}
